File: ios/Sources/AppCore/Views/CounterListPanel.swift
import SwiftUI

/// A simple scrollable list of every counter — colored dot + name + current
/// count — opened from the App Settings panel. Tapping a row selects that
/// counter and dismisses the list. Mirrors the panel chrome of
/// `CounterSettingsPanel`, and reuses `CounterTheme.dotColor` plus the blank-slot
/// em-dash / muted treatment from `CounterSwitcherCard`.
public struct CounterListPanel: View {
    let counters: [Counter]
    let activeId: Int
    let onSelect: (Int) -> Void
    let onClose: () -> Void

    public init(counters: [Counter],
                activeId: Int,
                onSelect: @escaping (Int) -> Void,
                onClose: @escaping () -> Void) {
        self.counters = counters
        self.activeId = activeId
        self.onSelect = onSelect
        self.onClose = onClose
    }

    private var sortedCounters: [Counter] {
        counters.sorted { $0.order < $1.order }
    }

    public var body: some View {
        VStack(alignment: .leading, spacing: 14) {
            header

            ScrollView {
                VStack(spacing: 0) {
                    ForEach(sortedCounters, id: \.id) { counter in
                        row(counter)
                        if counter.id != sortedCounters.last?.id {
                            Rectangle()
                                .fill(CounterTheme.line)
                                .frame(height: 1)
                        }
                    }
                }
            }
            .frame(maxHeight: 320)
        }
        .padding(20)
        .frame(maxWidth: .infinity, alignment: .top)
        .background(CounterTheme.panel)
        .overlay(Rectangle().stroke(CounterTheme.lineStrong, lineWidth: 1))
        .padding(.horizontal, 22)
        .padding(.top, 12)
    }

    private func row(_ counter: Counter) -> some View {
        let isBlank = counter.isBlank
        return Button(action: { onSelect(counter.id) }) {
            HStack(spacing: 12) {
                Circle()
                    .fill(isBlank ? CounterTheme.inkMuted : CounterTheme.dotColor(counter.colorKey))
                    .frame(width: 14, height: 14)
                    .overlay(
                        Circle()
                            .stroke(CounterTheme.accent, lineWidth: counter.id == activeId ? 2 : 0)
                            .padding(-3)
                    )
                Text(isBlank ? "—" : counter.name)
                    .font(.system(size: 16, weight: .heavy))
                    .foregroundColor(isBlank ? CounterTheme.inkMuted : CounterTheme.ink)
                    .lineLimit(1)
                Spacer()
                Text("\(counter.count)")
                    .font(.system(size: 15, weight: .bold, design: .monospaced))
                    .foregroundColor(CounterTheme.inkMuted)
            }
            .frame(maxWidth: .infinity)
            .padding(.vertical, 12)
            // Leading inset so the active dot's ring (a stroke that overhangs the
            // dot frame) isn't clipped by the scroll viewport's edge.
            .padding(.leading, 6)
            .contentShape(Rectangle())
        }
        .buttonStyle(.plain)
        .accessibilityIdentifier("counter-list-row-\(counter.id)")
    }

    private var header: some View {
        HStack {
            Text("ALL COUNTERS")
                .font(.system(size: 12, weight: .bold, design: .monospaced))
                .tracking(1.4)
                .foregroundColor(CounterTheme.inkMuted)
            Spacer()
            Button(action: onClose) {
                Text("DONE")
                    .font(.system(size: 12, weight: .heavy, design: .monospaced))
                    .tracking(1)
                    .foregroundColor(CounterTheme.onAccent)
                    .padding(.horizontal, 16)
                    .padding(.vertical, 8)
                    .background(CounterTheme.accent)
            }
            .buttonStyle(.plain)
            .accessibilityIdentifier("counter-list-close")
        }
    }
}

File: android/app/src/main/java/com/codeyam/android/ui/CounterListPanel.kt
package com.codeyam.android.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.RectangleShape
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.codeyam.android.model.Counter

/**
 * A scrollable list of every counter — colored dot, name, current count — opened
 * from the App Settings panel. Tapping a row selects that counter and dismisses
 * the list. Reuses the blank-slot em-dash / muted treatment from the switcher
 * card. Ports iOS `CounterListPanel`.
 */
@Composable
fun CounterListPanel(
    counters: List<Counter>,
    activeId: Int,
    onSelect: (Int) -> Unit,
    onClose: () -> Unit,
    modifier: Modifier = Modifier,
) {
    val sorted = counters.sortedBy { it.order }

    Column(
        modifier = modifier
            .fillMaxWidth()
            .padding(horizontal = 22.dp)
            .padding(top = 12.dp)
            .background(CounterColors.panel)
            .border(1.dp, CounterColors.lineStrong, RectangleShape)
            .padding(20.dp),
        verticalArrangement = Arrangement.spacedBy(14.dp),
    ) {
        PanelHeader(title = "ALL COUNTERS", doneIdentifier = "counter-list-close", onDone = onClose)

        // A FIXED height, not a max: iOS's `.frame(maxHeight: 320)` on a greedy
        // ScrollView always takes the full 320pt, so the card is tall enough to
        // fully cover the App Settings panel it was opened from. Hugging content
        // here instead would let that panel show through underneath.
        Column(
            modifier = Modifier
                .height(320.dp)
                .verticalScroll(rememberScrollState()),
        ) {
            sorted.forEachIndexed { index, counter ->
                CounterListRow(counter = counter, isActive = counter.id == activeId) {
                    onSelect(counter.id)
                }
                if (index != sorted.lastIndex) {
                    Box(
                        modifier = Modifier
                            .fillMaxWidth()
                            .height(1.dp)
                            .background(CounterColors.line),
                    )
                }
            }
        }
    }
}

@Composable
private fun CounterListRow(counter: Counter, isActive: Boolean, onClick: () -> Unit) {
    val interaction = remember { MutableInteractionSource() }
    val isBlank = counter.isBlank

    Row(
        modifier = Modifier
            .fillMaxWidth()
            .clickable(interactionSource = interaction, indication = null, onClick = onClick)
            .padding(vertical = 12.dp)
            // Leading inset so the active dot's ring isn't clipped by the scroll
            // viewport's edge.
            .padding(start = 6.dp)
            .semantics { contentDescription = "counter-list-row-${counter.id}" },
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(12.dp),
    ) {
        Box(
            modifier = Modifier
                .size(20.dp)
                .then(if (isActive) Modifier.border(2.dp, CounterColors.accent, CircleShape) else Modifier),
            contentAlignment = Alignment.Center,
        ) {
            Box(
                modifier = Modifier
                    .size(14.dp)
                    .clip(CircleShape)
                    .background(if (isBlank) CounterColors.inkMuted else dotColor(counter.colorKey)),
            )
        }
        Text(
            text = if (isBlank) "—" else counter.name,
            fontSize = 16.sp,
            fontWeight = FontWeight.Black,
            color = if (isBlank) CounterColors.inkMuted else CounterColors.ink,
            maxLines = 1,
            overflow = TextOverflow.Ellipsis,
            modifier = Modifier.weight(1f),
        )
        Text(
            text = formatCount(counter.count),
            fontSize = 15.sp,
            fontWeight = FontWeight.Bold,
            fontFamily = FontFamily.Monospace,
            color = CounterColors.inkMuted,
        )
    }
}
